src/audio/google/google_rtc_audio_processing.c

```diff
@@ -52,6 +52,31 @@ DECLARE_SOF_RT_UUID("google-rtc-audio-processing", google_rtc_audio_processing_u
 DECLARE_TR_CTX(google_rtc_audio_processing_tr, SOF_UUID(google_rtc_audio_processing_uuid),
 			   LOG_LEVEL_INFO);
 
+#if !(defined(__ZEPHYR__) && defined(CONFIG_XTENSA))
+/* Zephyr provides uncached memory for static variables on SMP, but we
+ * are single-core component and know we can safely use the cache for
+ * AEC work.  XTOS SOF is cached by default, so stub the Zephyr API.
+ */
+#define arch_xtensa_cached_ptr(p) (p)
+#endif
+
+#ifndef __ZEPHYR__
+#define ALWAYS_INLINE inline __attribute__((always_inline))
+#endif
+
+static __aligned(PLATFORM_DCACHE_ALIGN)
+uint8_t aec_mem_blob[CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_MEMORY_BUFFER_SIZE_BYTES];
+
+#define NUM_FRAMES (CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ \
+		    / GOOGLE_RTC_AUDIO_PROCESSING_FREQENCY_TO_PERIOD_FRAMES)
+#define CHAN_MAX CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_CHANNEL_MAX
+
+static __aligned(PLATFORM_DCACHE_ALIGN)
+float refoutbuf[CHAN_MAX][NUM_FRAMES];
+
+static __aligned(PLATFORM_DCACHE_ALIGN)
+float micbuf[CHAN_MAX][NUM_FRAMES];
```

Comment:
Heh, this is turning out more contentious than I thought it would be.  @cujomalainey didn't like it either.  AEC uses a **lot** of memory, the big one being a 200kb (soon to be larger) block.  Grabbing that all from the heap is problematic, it generally only works once in a single firmware instance, after which fragmentation prevents it from repeating. 

Maybe it's a culture thing.  My firmware background looks at this and says "You want that in the heap? that's crazy!"  But if there's consensus that it really needs to be there I can put back the dynamic allocation.

Reply:
you can do it if AEC is a loadable module, or the used huge memory will keep some memory banks active and can't be powered off to save power.

Reply:
Loadable modules have the same fragmentation problems though (at least until an MMU arrives).  Also IIRC loadable modules get their own .data/.bss sections anyway and would be able to duplicate this without having to hit the heap (though again, likely not successfully on current hardware!).   It might also be possible to put this in IMR, but that would require some benchmarking.

Really my position here is: on all current deployments this module uses memory that can only be successfully be allocated at firmware init time.   And if that's the case, it should be static.

Reply:
I meant if it is loadable, the memory can be released when the module is unload for AEC is inactive. But for build-in module, the memory will be active even the AEC pipeline is deleted.

Reply:
Yes, that's the feature being provided here.  If you try to delete a 300kb+ block in a live heap that is only about twice that big, you're probably never going to get it back again. :)  Dynamic allocation of objects this large just doesn't work well.